### PR TITLE
Include output stream in `panic!()` documentation

### LINKED ISSUE #83252: std::panic macro documentation does not specify where the string is printed to (stdout or stderr)
Looking at the documentation it's not clear whether the string argument is printed to stdout or stderr.

### PR DESCRIPTION
**Pull request by jfrimmel**
Fixes #83252.

### CODE REVIEW

**Comment by m-ou-se** [comment]
> ``The string built by this macro (using the [`format!`] syntax for building the actual message) is printed to `stderr`.``

This is only what the default hook in `std` does, which can be overridden, or might not even exist (in `no_std` programs). The panic macro itself doesn't print anything to `stderr`; the concept of `stderr` or printing doesn't even existing `core`.

I'm not objecting to documenting what happens by default with `std`, but we shouldn't claim that this behaviour is a property of the `panic!()` macro itself. Maybe something like *"std's default panic hook will .."*, with a link to `std::panic::set_hook`.

@bors r-

**Comment by jfrimmel** [comment]
I've incorporated your feedback @m-ou-se. I hope this is both helpful to a reader while still being technically correct. Despite this file being in `core`, I'd like to mention the `std`-behavior prominently, as most of the readers come from reading the `std` docs.

File: library/core/src/macros/panic.md [hunk 1]
```diff
@@ -9,11 +9,14 @@ tests. `panic!` is closely tied with the `unwrap` method of both
...
-This macro is used to inject panic into a Rust thread, causing the thread to
-panic entirely. This macro panics with a string and uses the [`format!`] syntax
-for building the message.
-
-Each thread's panic can be reaped as the [`Box`]`<`[`Any`]`>` type,
+When using `panic!()` you can specify a string payload, that is built using
+the [`format!`] syntax. That payload is used when injecting the panic into
+the calling Rust thread, causing the thread to panic entirely.
+
+The default `std` panic handling strategy is to print the message payload to
```
Comment (m-ou-se) [inline library/core/src/macros/panic.md:16]: The panic *handling* stragety of `std` is to unwind or abort. Printing to stderr is done as part of the panic *hook*, which executes before unwinding/aborting happens. Even if you handle a panic yourself with `catch_unwind` or `thread.join()`, the hook was already executed and already printed the panic to stderr.

File: library/core/src/macros/panic.md [hunk 2]
```diff
@@ -9,11 +9,14 @@ tests. `panic!` is closely tied with the `unwrap` method of both
...
+the `stderr` along with the file/line/column information of the `panic!()`
+call. You can override the panic hook using [`std::panic::set_hook()`].
+Inside the hook a panic can be reaped as the [`Box`]`<`[`Any`]`>` type,
```
Comment (m-ou-se) [inline library/core/src/macros/panic.md:19]: You don't get access to the `Box<Any>` in the hook. You can only look at the payload through [`PanicInfo::payload()`](https://doc.rust-lang.org/stable/std/panic/struct.PanicInfo.html#method.payload), which gives a `&(dyn Any + Send)`. Only when catching a panic (with `catch_unwind` or `join`ing a thread) you get the Box.

**Review by m-ou-se — COMMENTED** [review]
Thanks. Looking better now. I have a few more comments:

File: library/core/src/macros/panic.md (hunk 1 above)
Comment (jfrimmel) [reply library/core/src/macros/panic.md]: Okay, I've chosen the wrong terminology. Should be fixed.

File: library/core/src/macros/panic.md (hunk 2 above)
Comment (jfrimmel) [reply library/core/src/macros/panic.md]: Oh, I was mislead by the existing documentation. The proposed change does not mention the `Box<Any>` anymore, as this does not integrate well into the rest of the paragraph and probably should be in the docs of `panic::catch_unwind()`].

**Comment by jfrimmel** [comment]
Didn't intend to make this much noise in such a small change, apologies.

**Comment by m-ou-se** [comment]
> Didn't intend to make this much noise in such a small change, apologies.

Oh I'm very happy you're doing this! Clearly we were missing documentation here. Looks like we didn't even explain anywhere what happens with `panic!()` in `std`. So this is an important change you're making!

**Comment by m-ou-se** [comment]
This looks great now. Thanks!

@bors r+ rollup

File: library/core/src/macros/panic.md [hunk 3]
```diff
@@ -9,11 +9,15 @@ tests. `panic!` is closely tied with the `unwrap` method of both
...
-This macro is used to inject panic into a Rust thread, causing the thread to
-panic entirely. This macro panics with a string and uses the [`format!`] syntax
-for building the message.
-
-Each thread's panic can be reaped as the [`Box`]`<`[`Any`]`>` type,
+When using `panic!()` you can specify a string payload, that is built using
+the [`format!`] syntax. That payload is used when injecting the panic into
+the calling Rust thread, causing the thread to panic entirely.
+
+The behavior of the default `std` hook, i.e. the code, that runs directy
```
Comment (joshtriplett) [inline library/core/src/macros/panic.md:16]:
```suggestion
The behavior of the default `std` hook, i.e. the code that runs directly
```

File: library/core/src/macros/panic.md [hunk 4]
```diff
@@ -9,11 +9,15 @@ tests. `panic!` is closely tied with the `unwrap` method of both
...
+after the panic is invoked, is to print the message payload to the
```
Comment (joshtriplett) [inline library/core/src/macros/panic.md:17]:
```suggestion
after the panic is invoked, is to print the message payload to
```

File: library/core/src/macros/panic.md [hunk 5]
```diff
@@ -9,11 +9,15 @@ tests. `panic!` is closely tied with the `unwrap` method of both
...
+`stderr` along with the file/line/column information of the `panic!()`
+call. You can override the panic hook using [`std::panic::set_hook()`].
+Inside the hook a panic can be accessed as a `&dyn Any + Send`, which
```
Comment (joshtriplett) [inline library/core/src/macros/panic.md:20]:
```suggestion
Inside the hook a panic can be accessed as a `&dyn Any + Send`,
```

**Review by joshtriplett — CHANGES_REQUESTED** [review]
A few typos; otherwise LGTM, and I'm glad to see this documented.

**Comment by joshtriplett** [comment]
Thanks for the quick update! :+1:

**Pull request merged**